File: biglots-scraper/browser/python/biglots.py
"""Big Lots scraper using the official Scrapeless Python SDK + Playwright over CDP.

Surfaces:
- scrape_product(urls)                  -> list of Product dicts (PDP fields lifted from JSON-LD)
- scrape_search(category_url, max_pages) -> list of SearchResult dicts (WooCommerce cards)

biglots.com runs WordPress + WooCommerce; JSON-LD Product blocks are emitted on every PDP,
and the category templates render `<li class="wp-block-post post-<id> product ...">` cards.
"""

from __future__ import annotations

import json
import os
import re
from typing import Any, Dict, List, Optional

from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 180

# ---------------------------------------------------------------------------
# Scrapeless plumbing
# ---------------------------------------------------------------------------

def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()

async def _fetch_rendered_html(
    url: str,
    ready_selector: Optional[str] = None,
    *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY,
    retries: int = 1,
) -> str:
    last_error: Optional[Exception] = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
        )
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(session.browser_ws_endpoint)
            try:
                page = await browser.new_page()
                await page.goto(url, wait_until="domcontentloaded", timeout=45000)
                if ready_selector:
                    try:
                        await page.wait_for_selector(ready_selector, timeout=15000)
                    except Exception as e:
                        logger.warning("wait_for_selector {!r} failed (continuing): {}", ready_selector, e)
                html = await page.content()
                if html and "<html" in html.lower():
                    return html
                last_error = RuntimeError("empty HTML")
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("attempt {} failed: {}", attempt + 1, e)
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
    raise RuntimeError(f"giving up after {retries + 1} attempts: {last_error}")

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _clean(value: Any) -> Optional[str]:
    if value is None:
        return None
    s = re.sub(r"\s+", " ", str(value)).strip()
    return s or None

def _to_number(value: Any) -> Optional[float]:
    if value is None:
        return None
    s = re.sub(r"[^\d.\-]", "", str(value))
    try:
        return float(s)
    except (TypeError, ValueError):
        return None

def _iter_jsonld_nodes(sel: Selector):
    for raw in sel.css('script[type="application/ld+json"]::text').getall():
        if not raw or not raw.strip():
            continue
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            continue
        stack: List[Any] = list(data) if isinstance(data, list) else [data]
        while stack:
            node = stack.pop(0)
            if not isinstance(node, dict):
                continue
            graph = node.get("@graph")
            if isinstance(graph, list):
                stack.extend(graph)
            else:
                yield node

def _type_matches(node: Dict[str, Any], wanted: str) -> bool:
    t = node.get("@type")
    if isinstance(t, str):
        return t == wanted
    if isinstance(t, list):
        return wanted in t
    return False

def _pick_offer_price(offer: Dict[str, Any]) -> Optional[float]:
    if not isinstance(offer, dict):
        return None
    if offer.get("price") is not None:
        return _to_number(offer["price"])
    spec = offer.get("priceSpecification")
    if isinstance(spec, list) and spec and isinstance(spec[0], dict):
        return _to_number(spec[0].get("price"))
    if isinstance(spec, dict):
        return _to_number(spec.get("price"))
    return None

def _pick_offer_currency(offer: Dict[str, Any]) -> Optional[str]:
    if not isinstance(offer, dict):
        return None
    if offer.get("priceCurrency"):
        return str(offer["priceCurrency"])
    spec = offer.get("priceSpecification")
    if isinstance(spec, list) and spec and isinstance(spec[0], dict):
        return spec[0].get("priceCurrency")
    if isinstance(spec, dict):
        return spec.get("priceCurrency")
    return None

def _short_availability(value: Any) -> Optional[str]:
    if not isinstance(value, str):
        return None
    return value.rsplit("/", 1)[-1] if "/" in value else value

# ---------------------------------------------------------------------------
# Parsers
# ---------------------------------------------------------------------------

def parse_product(html: str, url: str) -> Dict[str, Any]:
    sel = Selector(text=html)

    ld: Dict[str, Any] = {}
    for node in _iter_jsonld_nodes(sel):
        if _type_matches(node, "Product"):
            ld = node
            break

    offers = ld.get("offers")
    if isinstance(offers, list) and offers:
        offer = offers[0] if isinstance(offers[0], dict) else {}
    elif isinstance(offers, dict):
        offer = offers
    else:
        offer = {}

    images_raw = ld.get("image")
    if isinstance(images_raw, str):
        images = [images_raw]
    elif isinstance(images_raw, list):
        images = [str(x) for x in images_raw if x]
    else:
        images = sel.css('meta[property="og:image"]::attr(content)').getall()

    categories: List[str] = []
    for t in sel.css('.woocommerce-breadcrumb a::text, .wp-block-woocommerce-breadcrumbs a::text').getall():
        c = _clean(t)
        if c and c not in categories:
            categories.append(c)

    seller_node = offer.get("seller")
    seller_name: Optional[str] = None
    if isinstance(seller_node, dict):
        seller_name = seller_node.get("name")
    elif isinstance(seller_node, str):
        seller_name = seller_node

    sku = ld.get("sku")
    sku_str = str(sku) if sku is not None else (ld.get("@id") or url)

    return {
        "id": str(sku_str),
        "url": url,
        "name": _clean(ld.get("name")) or _clean(sel.css("h1::text").get()) or "",
        "description": _clean(ld.get("description")) or _clean(sel.css('meta[name="description"]::attr(content)').get()),
        "price": _pick_offer_price(offer),
        "priceCurrency": _pick_offer_currency(offer),
        "availability": _short_availability(offer.get("availability")),
        "images": images,
        "categories": categories,
        "sellerName": _clean(seller_name),
    }

_POST_CLASS_RE = re.compile(r"^post-(\d+)$")

def parse_search(html: str) -> List[Dict[str, Any]]:
    sel = Selector(text=html)
    items: List[Dict[str, Any]] = []
    seen: set[str] = set()

    for card in sel.css('li.product, li.wp-block-post.product'):
        classes = (card.attrib.get("class") or "").split()
        post_id = ""
        for c in classes:
            m = _POST_CLASS_RE.match(c)
            if m:
                post_id = m.group(1)
                break
        card_id = post_id or card.attrib.get("data-id", "")

        href = card.css('a[href*="/product/"]::attr(href)').get() or ""
        if not href:
            continue
        abs_url = href if href.startswith("http") else f"https://biglots.com{href}"
        if abs_url in seen:
            continue
        seen.add(abs_url)

        name = _clean(
            card.css('h3 a::text, h2 a::text, h3::text, h2::text').get()
            or card.css('a[href*="/product/"]::attr(aria-label)').get()
            or card.css('img::attr(alt)').get()
        ) or ""

        price_text = _clean(
            " ".join(card.css(
                '.wp-block-woocommerce-product-price *::text, '
                '.wc-block-components-product-price *::text, '
                '.price *::text'
            ).getall())
        )
        price = _to_number(price_text)

        image = card.css('img::attr(src)').get() or card.css('img::attr(data-src)').get()

        category = _clean(card.css('.wp-block-post-terms a::text, .wc-block-grid__product-category::text').get())

        items.append({
            "id": str(card_id or ""),
            "url": abs_url,
            "name": name,
            "image": image,
            "price": price,
            "priceCurrency": "USD" if price_text and "$" in price_text else None,
            "category": category,
        })

    return items

# ---------------------------------------------------------------------------
# Scrape functions
# ---------------------------------------------------------------------------

async def scrape_product(urls: List[str]) -> List[Dict[str, Any]]:
    if isinstance(urls, str):
        urls = [urls]
    out: List[Dict[str, Any]] = []
    for url in urls:
        logger.info("scraping product {}", url)
        html = await _fetch_rendered_html(url, ready_selector='script[type="application/ld+json"]')
        out.append(parse_product(html, url))
    return out

async def scrape_search(category_url: str, max_pages: int = 1) -> List[Dict[str, Any]]:
    out: List[Dict[str, Any]] = []
    base = category_url.rstrip("/")
    for page in range(1, max_pages + 1):
        url = category_url if page == 1 else f"{base}/page/{page}/"
        logger.info("scraping search {} (page {})", url, page)
        html = await _fetch_rendered_html(url, ready_selector='li.product, li.wp-block-post.product')
        out.extend(parse_search(html))
    logger.success("search returned {} items", len(out))
    return out

def to_dict(obj: Any) -> Any:
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj

File: biglots-scraper/browser/nodejs/biglots.mjs
// Big Lots scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// Surfaces:
//   - scrapeProduct(urls)                  -> array of Product dicts (PDP fields lifted from JSON-LD)
//   - scrapeSearch(categoryUrl, maxPages)  -> array of SearchResult dicts (WooCommerce cards)
//
// biglots.com runs WordPress + WooCommerce. JSON-LD Product blocks are emitted on
// every PDP; the category pages render WooCommerce block-template <li class="post-...">
// cards, so DOM scraping is straightforward.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 180;

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function fetchRenderedHtml(url, readySelector, { proxyCountry = DEFAULT_PROXY_COUNTRY, retries = 1 } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const { browserWSEndpoint } = await client().browser.create({
      proxyCountry,
      sessionTTL: DEFAULT_SESSION_TTL,
    });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint });
      const page = await browser.newPage();
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 45000 });
      if (readySelector) {
        try { await page.waitForSelector(readySelector, { timeout: 15000 }); } catch (_) {}
      }
      const html = await page.content();
      if (html && html.toLowerCase().includes("<html")) return html;
      lastError = new Error("empty HTML");
    } catch (e) {
      lastError = e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
  }
  throw new Error(`giving up after ${retries + 1} attempts: ${lastError?.message}`);
}

// ---------------- helpers ----------------

function clean(value) {
  if (value == null) return null;
  const v = String(value).replace(/\s+/g, " ").trim();
  return v || null;
}

function toNumber(value) {
  if (value == null) return null;
  const n = parseFloat(String(value).replace(/[^\d.\-]/g, ""));
  return Number.isFinite(n) ? n : null;
}

function* iterJsonldNodes($) {
  const blocks = $('script[type="application/ld+json"]')
    .map((_, el) => $(el).contents().text())
    .get();
  for (const raw of blocks) {
    if (!raw || !raw.trim()) continue;
    let data;
    try { data = JSON.parse(raw); } catch { continue; }
    const stack = Array.isArray(data) ? [...data] : [data];
    while (stack.length) {
      const node = stack.shift();
      if (!node || typeof node !== "object") continue;
      if (Array.isArray(node["@graph"])) {
        for (const sub of node["@graph"]) stack.push(sub);
      } else {
        yield node;
      }
    }
  }
}

function typeMatches(node, wanted) {
  const t = node["@type"];
  if (typeof t === "string") return t === wanted;
  if (Array.isArray(t)) return t.includes(wanted);
  return false;
}

function pickOfferPrice(offer) {
  if (!offer || typeof offer !== "object") return null;
  if (offer.price != null) return toNumber(offer.price);
  const spec = offer.priceSpecification;
  if (Array.isArray(spec) && spec.length) return toNumber(spec[0].price);
  if (spec && typeof spec === "object") return toNumber(spec.price);
  return null;
}

function pickOfferCurrency(offer) {
  if (!offer || typeof offer !== "object") return null;
  if (offer.priceCurrency) return String(offer.priceCurrency);
  const spec = offer.priceSpecification;
  if (Array.isArray(spec) && spec.length) return spec[0].priceCurrency ?? null;
  if (spec && typeof spec === "object") return spec.priceCurrency ?? null;
  return null;
}

function shortAvailability(value) {
  if (typeof value !== "string") return null;
  return value.includes("/") ? value.split("/").pop() : value;
}

// ---------------- parsers ----------------

export function parseProduct(html, url) {
  const $ = cheerio.load(html);

  let ld = {};
  for (const node of iterJsonldNodes($)) {
    if (typeMatches(node, "Product")) { ld = node; break; }
  }

  const offers = ld.offers;
  let offer = {};
  if (Array.isArray(offers) && offers.length) offer = offers[0];
  else if (offers && typeof offers === "object") offer = offers;

  let images = [];
  const imgRaw = ld.image;
  if (typeof imgRaw === "string") images = [imgRaw];
  else if (Array.isArray(imgRaw)) images = imgRaw.filter(Boolean).map(String);
  if (!images.length) {
    images = $('meta[property="og:image"]').map((_, el) => $(el).attr("content")).get().filter(Boolean);
  }

  const categories = [];
  $(".woocommerce-breadcrumb a, .wp-block-woocommerce-breadcrumbs a").each((_, el) => {
    const t = clean($(el).text());
    if (t && !categories.includes(t)) categories.push(t);
  });

  const sellerNode = offer.seller;
  let sellerName = null;
  if (sellerNode && typeof sellerNode === "object") sellerName = sellerNode.name ?? null;
  else if (typeof sellerNode === "string") sellerName = sellerNode;

  const sku = ld.sku != null ? String(ld.sku) : (ld["@id"] || url);

  return {
    id: String(sku),
    url,
    name: clean(ld.name) || clean($("h1").first().text()) || "",
    description: clean(ld.description) || clean($('meta[name="description"]').attr("content")),
    price: pickOfferPrice(offer),
    priceCurrency: pickOfferCurrency(offer),
    availability: shortAvailability(offer.availability),
    images,
    categories,
    sellerName: clean(sellerName),
  };
}

export function parseSearch(html) {
  const $ = cheerio.load(html);
  const items = [];
  const seen = new Set();

  $('li.product, li.wp-block-post.product').each((_, el) => {
    const $card = $(el);
    // post-<id> class identifies the WP post id; fall back to data-id if present
    const classes = ($card.attr("class") || "").split(/\s+/);
    const postClass = classes.find((c) => /^post-\d+$/.test(c));
    const id = postClass ? postClass.replace("post-", "") : ($card.attr("data-id") ?? "");

    const anchor = $card.find('a[href*="/product/"]').first();
    const href = anchor.attr("href") || "";
    if (!href) return;
    const url = href.startsWith("http") ? href : `https://biglots.com${href}`;
    if (seen.has(url)) return;
    seen.add(url);

    const name = clean(
      $card.find('h3 a, h2 a, h3, h2').first().text() ||
      anchor.attr("aria-label") ||
      $card.find("img").first().attr("alt")
    ) || "";

    const priceText = clean(
      $card.find('.wp-block-woocommerce-product-price, .wc-block-components-product-price, .price').first().text()
    );
    const price = toNumber(priceText);

    const image = $card.find('img').first().attr('src') || $card.find('img').first().attr('data-src') || null;

    const category = clean(
      $card.find('.wp-block-post-terms a, .wc-block-grid__product-category').first().text()
    );

    items.push({
      id: String(id || ""),
      url,
      name,
      image: image || null,
      price,
      priceCurrency: priceText && priceText.includes("$") ? "USD" : null,
      category,
    });
  });

  return items;
}

// ---------------- scrape functions ----------------

export async function scrapeProduct(urls) {
  const list = Array.isArray(urls) ? urls : [urls];
  const out = [];
  for (const url of list) {
    const html = await fetchRenderedHtml(url, 'script[type="application/ld+json"]');
    out.push(parseProduct(html, url));
  }
  return out;
}

export async function scrapeSearch(categoryUrl, maxPages = 1) {
  const out = [];
  for (let page = 1; page <= maxPages; page++) {
    let url = categoryUrl;
    if (page > 1) {
      // WP archive pagination: <category>/page/<n>/
      const trimmed = categoryUrl.replace(/\/$/, "");
      url = `${trimmed}/page/${page}/`;
    }
    const html = await fetchRenderedHtml(url, 'li.product, li.wp-block-post.product');
    out.push(...parseSearch(html));
  }
  return out;
}
